larpandoracontent/LArTwoDReco/HitWidthClusterMergingAlgorithm.cc

```diff
@@ -0,0 +1,503 @@
+
+        const CartesianVector midpoint((currentFitParameters.GetHigherXExtrema() + testFitParameters.GetLowerXExtrema())*0.5);
+        
+        CartesianVector newClusterDirection(GetClusterDirection(newParameters, midpoint));
+        
+        if(std::fabs(newClusterDirection.GetCosOpeningAngle(currentClusterDirection)) < m_maxDirectionDeviationCosAngle || std::fabs(newClusterDirection.GetCosOpeningAngle(testClusterDirection)) < m_maxDirectionDeviationCosAngle)
+            return false;
+    }
+    
+    return true;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool HitWidthClusterMergingAlgorithm::IsExtremalCluster(const bool isForward, const Cluster *const pCurrentCluster,  const Cluster *const pTestCluster) const
+{
+
+    LArHitWidthHelper::ClusterToParametersMapStore* pClusterToParametersMapStore = LArHitWidthHelper::ClusterToParametersMapStore::Instance();
+    LArHitWidthHelper::ClusterToParametersMap* pClusterToParametersMap = pClusterToParametersMapStore->GetMap();
+
+    if(pClusterToParametersMap->empty())
+        throw StatusCodeException(STATUS_CODE_NOT_INITIALIZED);
+
+    LArHitWidthHelper::ClusterToParametersMap::const_iterator currentIter(pClusterToParametersMap->find(pCurrentCluster));
+    LArHitWidthHelper::ClusterToParametersMap::const_iterator testIter(pClusterToParametersMap->find(pTestCluster));
+
+    if(currentIter == pClusterToParametersMap->end() || testIter == pClusterToParametersMap->end())
+        throw StatusCodeException(STATUS_CODE_NOT_FOUND);
+
+    CartesianVector currentHigherXExtrema(currentIter->second.GetHigherXExtrema()), testHigherXExtrema(testIter->second.GetHigherXExtrema());
+
+    float currentMaxX(currentHigherXExtrema.GetX()), testMaxX(testHigherXExtrema.GetX());
+
+    if (isForward)
+    {
+        if (std::fabs(testMaxX - currentMaxX) > std::numeric_limits<float>::epsilon())
+            return (testMaxX > currentMaxX);
+    }
+    else
+    {
+        if (std::fabs(testMaxX - currentMaxX) > std::numeric_limits<float>::epsilon())
+            return (testMaxX < currentMaxX);
+    }
+  
+    return LArClusterHelper::SortByNHits(pTestCluster, pCurrentCluster);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+CartesianVector HitWidthClusterMergingAlgorithm::GetClusterDirection(const LArHitWidthHelper::ClusterParameters &clusterFitParameters, const CartesianVector &fitReferencePoint) const 
+{
+
+    // If cluster composed of one hit, return a transverse line as fit
+    if(clusterFitParameters.GetNumCaloHits() == 1)
+        return CartesianVector(1, 0, 0);
+
+    // minimise the longitudinal distance in fit (works best for transverse fits)
+    CartesianVector LSTransverseClusterFitDirection(0,0,0);
```

Comment:
Conventionally start variable names with lower case, leaving upper case names for types.